```
Error running mint automation
+ mkdir -p 16279-d562491/gopath/src/github.com/minio
+ git -C 16279-d562491/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 16279-d562491/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 16279-d562491/gopath/src/github.com/minio/minio fetch --quiet minio pull/16279/head:pr16279
+ git -C 16279-d562491/gopath/src/github.com/minio/minio checkout --quiet pr16279
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/16279-d562491/gopath
+ make -C 16279-d562491/gopath/src/github.com/minio/minio --quiet
go: downloading cloud.google.com/go/storage v1.26.0
go: downloading github.com/djherbis/atime v1.1.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/Azure/azure-storage-blob-go v0.15.0
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/minio/kes v0.22.0
go: downloading github.com/miekg/dns v1.1.50
go: downloading github.com/minio/minio-go/v7 v7.0.44
go: downloading github.com/bits-and-blooms/bloom/v3 v3.3.1
go: downloading github.com/klauspost/compress v1.15.12
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/coreos/go-oidc v2.2.1+incompatible
go: downloading github.com/minio/pkg v1.5.8
go: downloading github.com/minio/console v0.22.2
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/klauspost/filepathx v1.1.1
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/minio/zipindex v0.3.0
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/minio/cli v1.24.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/lithammer/shortuuid/v4 v4.0.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/go-openapi/loads v0.21.2
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/felixge/fgprof v0.9.3
go: downloading github.com/klauspost/reedsolomon v1.11.3
go: downloading github.com/dchest/siphash v1.2.3
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/inconshreveable/mousetrap v1.0.1
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading google.golang.org/api v0.98.0
go: downloading github.com/golang-jwt/jwt/v4 v4.4.2
go: downloading golang.org/x/oauth2 v0.2.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading golang.org/x/sys v0.2.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.5
go: downloading golang.org/x/crypto v0.3.0
go: downloading go.etcd.io/etcd/client/v3 v3.5.5
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/tinylib/msgp v1.1.7-0.20220719154719-f3635b96e483
go: downloading github.com/prometheus/client_golang v1.13.0
go: downloading github.com/shirou/gopsutil/v3 v3.22.9
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/prometheus/client_model v0.3.0
go: downloading github.com/rs/cors v1.8.2
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading go.uber.org/atomic v1.10.0
go: downloading github.com/prometheus/common v0.37.0
go: downloading github.com/prometheus/procfs v0.8.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading go.uber.org/zap v1.23.0
go: downloading github.com/coredns/coredns v1.9.4
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/go-ldap/ldap/v3 v3.4.4
go: downloading golang.org/x/time v0.0.0-20220722155302-e5dcc9cfc0b9
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/Shopify/sarama v1.36.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.7
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.9
go: downloading github.com/lib/pq v1.10.7
go: downloading github.com/nats-io/nats.go v1.17.0
go: downloading github.com/nats-io/stan.go v0.10.3
go: downloading github.com/nsqio/go-nsq v1.1.0
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/rabbitmq/amqp091-go v1.5.0
go: downloading github.com/xdg/scram v1.0.5
go: downloading golang.org/x/sync v0.1.0
go: downloading github.com/bits-and-blooms/bitset v1.3.3
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/mattn/go-isatty v0.0.16
go: downloading github.com/mattn/go-runewidth v0.0.14
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.6.0
go: downloading github.com/minio/csvparser v1.0.0
go: downloading golang.org/x/net v0.2.0
go: downloading github.com/fraugster/parquet-go v0.12.0
go: downloading github.com/alecthomas/participle v0.7.1
go: downloading github.com/google/pprof v0.0.0-20220829040838-70bd9ae97f40
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/klauspost/cpuid/v2 v2.2.1
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.5
go: downloading google.golang.org/grpc v1.50.1
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/xdg/stringprep v1.0.3
go: downloading github.com/rivo/uniseg v0.4.2
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading google.golang.org/genproto v0.0.0-20221018160656-63c7b68cfc55
go: downloading cloud.google.com/go v0.104.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.3
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/go-openapi/analysis v0.21.4
go: downloading github.com/go-openapi/spec v0.20.7
go: downloading github.com/go-openapi/swag v0.22.3
go: downloading google.golang.org/protobuf v1.28.1
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/rs/xid v1.4.0
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading gopkg.in/ini.v1 v1.67.0
go: downloading golang.org/x/text v0.4.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/lestrrat-go/jwx v1.2.25
go: downloading aead.dev/mem v0.2.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/coreos/go-systemd/v22 v22.4.0
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.3
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/jcmturner/gofork v1.7.6
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/pierrec/lz4/v4 v4.1.16
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/apache/thrift v0.16.0
go: downloading github.com/go-openapi/strfmt v0.21.3
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/blackmagic v1.0.1
go: downloading github.com/lestrrat-go/iter v1.0.2
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading github.com/go-openapi/errors v0.20.3
go: downloading go.mongodb.org/mongo-driver v1.10.3
go: downloading github.com/oklog/ulid v1.3.1
go: downloading cloud.google.com/go/iam v0.4.0
go: downloading cloud.google.com/go/compute v1.10.0
go: downloading github.com/googleapis/gax-go/v2 v2.5.1
go: downloading go.opencensus.io v0.23.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-openapi/runtime v0.24.2
go: downloading github.com/go-openapi/validate v0.22.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/unrolled/secure v1.13.0
go: downloading github.com/minio/mc v0.0.0-20221201184114-854b4f123f03
go: downloading github.com/tidwall/gjson v1.14.3
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/tidwall/pretty v1.2.1
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/jedib0t/go-pretty/v6 v6.3.8
go: downloading github.com/charmbracelet/bubbles v0.14.0
go: downloading github.com/navidys/tvxwidgets v0.1.1
go: downloading github.com/minio/colorjson v1.0.4
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/tklauser/go-sysconf v0.3.11
go: downloading golang.org/x/term v0.2.0
go: downloading github.com/posener/complete v1.2.3
go: downloading github.com/charmbracelet/lipgloss v0.6.0
go: downloading github.com/charmbracelet/bubbletea v0.22.1
go: downloading github.com/gdamore/tcell/v2 v2.5.3
go: downloading github.com/pkg/xattr v0.4.9
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/prometheus/prom2json v1.3.2
go: downloading github.com/rivo/tview v0.0.0-20220916081518-2e69b7385a37
go: downloading github.com/muesli/termenv v0.13.0
go: downloading github.com/muesli/cancelreader v0.2.2
go: downloading github.com/mattn/go-localereader v0.0.1
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/tklauser/numcpus v0.6.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading github.com/aymanbagabas/go-osc52 v1.2.1
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.1.0
go: downloading github.com/google/go-cmp v0.5.9
../../../../pkg/mod/github.com/minio/console@v0.22.2/pkg/logger/config/config.go:20:2: github.com/minio/madmin-go/v2@v2.0.1: replacement directory ../madmin-go does not exist
../../../../pkg/mod/github.com/minio/pkg@v1.5.8/sys/stats_linux.go:27:2: github.com/minio/madmin-go/v2@v2.0.1: replacement directory ../madmin-go does not exist
cmd/admin-handlers.go:51:2: github.com/minio/madmin-go/v2@v2.0.1: replacement directory ../madmin-go does not exist
cmd/server-rlimit.go:24:2: github.com/minio/madmin-go/v2@v2.0.1: replacement directory ../madmin-go does not exist
make: *** [build] Error 1

```

--- a/cmd/api-errors.go
+++ b/cmd/api-errors.go
@@ -1384,7 +1384,7 @@ var errorCodes = errorCodeMap{
 	ErrSiteReplicationPeerResp: {
 		Code:           "XMinioSiteReplicationPeerResp",
 		Description:    "Error received when contacting a peer site",
-		HTTPStatusCode: http.StatusServiceUnavailable,
+		HTTPStatusCode: http.StatusBadRequest,
 	},
 	ErrSiteReplicationBackendIssue: {
 		Code:           "XMinioSiteReplicationBackendIssue",

--- a/cmd/bucket-replication.go
+++ b/cmd/bucket-replication.go
@@ -131,8 +131,8 @@ func validateReplicationDestination(ctx context.Context, bucket string, rCfg *re
 			}
 		}
 		// validate replication ARN against target endpoint
-		c, ok := globalBucketTargetSys.arnRemotesMap[arnStr]
-		if ok {
+		c := globalBucketTargetSys.GetRemoteTargetClient(ctx, arnStr)
+		if c != nil {
 			if c.EndpointURL().String() == clnt.EndpointURL().String() {
 				selfTarget, _ := isLocalHost(clnt.EndpointURL().Hostname(), clnt.EndpointURL().Port(), globalMinioPort)
 				if !sameTarget {

--- a/cmd/site-replication.go
+++ b/cmd/site-replication.go
@@ -72,6 +72,10 @@ var (
 		Cause: errors.New("peer not found"),
 		Code:  ErrSiteReplicationInvalidRequest,
 	}
+	errSRRequestorNotFound = SRError{
+		Cause: errors.New("requesting site not found in site replication config"),
+		Code:  ErrSiteReplicationInvalidRequest,
+	}
 	errSRNotEnabled = SRError{
 		Cause: errors.New("site replication is not enabled"),
 		Code:  ErrSiteReplicationInvalidRequest,
@@ -2060,12 +2064,12 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 		}
 	}
 	for _, s := range siteNames {
-		info, ok := peerMap[s]
+		pinfo, ok := peerMap[s]
 		if !ok {
 			return st, errSRConfigMissingError(errMissingSRConfig)
 		}
-		rmvEndpoints = append(rmvEndpoints, info.Endpoint)
-		delete(updatedPeers, info.DeploymentID)
+		rmvEndpoints = append(rmvEndpoints, pinfo.Endpoint)
+		delete(updatedPeers, pinfo.DeploymentID)
 	}
 	var wg sync.WaitGroup
 	errs := make(map[string]error, len(c.state.Peers))
@@ -2087,6 +2091,8 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 				errs[pi.DeploymentID] = errSRPeerResp(fmt.Errorf("unable to create admin client for %s: %w", pi.Name, err))
 				return
 			}
+			// set the requesting site's deploymentID for verification of peer request
+			rreq.RequestingDepID = globalDeploymentID
 			if _, err = admClient.SRPeerRemove(ctx, rreq); err != nil {
 				if errors.Is(err, errMissingSRConfig) {
 					// ignore if peer is already removed.
@@ -2100,9 +2106,11 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 	wg.Wait()
 
 	errdID := ""
+	selfTgtsDeleted := errs[globalDeploymentID] == nil // true if all remote targets and replication config cleared successfully on local cluster
+
 	for dID, err := range errs {
 		if err != nil {
-			if !rreq.RemoveAll {
+			if !rreq.RemoveAll && !selfTgtsDeleted {
 				return madmin.ReplicateRemoveStatus{
 					ErrDetail: err.Error(),
 					Status:    madmin.ReplicateRemoveStatusPartial,
@@ -2111,7 +2119,10 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 			errdID = dID
 		}
 	}
-
+	var remoteErr error // make sure to return err from remote if any
+	if errdID != "" {
+		remoteErr = errs[errdID]
+	}
 	// force local config to be cleared even if peers failed since the remote targets are deleted
 	// by now from the replication config and user intended to forcibly clear all site replication
 	if rreq.RemoveAll {
@@ -2146,12 +2157,16 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 		return madmin.ReplicateRemoveStatus{
 			Status:    madmin.ReplicateRemoveStatusPartial,
 			ErrDetail: fmt.Sprintf("unable to save cluster-replication state on local: %v", err),
-		}, nil
+		}, remoteErr
 	}
 
-	return madmin.ReplicateRemoveStatus{
+	st = madmin.ReplicateRemoveStatus{
 		Status: madmin.ReplicateRemoveStatusSuccess,
-	}, nil
+	}
+	if remoteErr != nil {
+		st.Status = madmin.ReplicateRemoveStatusPartial
+	}
+	return st, remoteErr
 }
 
 // InternalRemoveReq - sends an unlink request to peer cluster to remove one or more sites
@@ -2159,6 +2174,19 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 func (c *SiteReplicationSys) InternalRemoveReq(ctx context.Context, objectAPI ObjectLayer, rreq madmin.SRRemoveReq) error {
 	if !c.isEnabled() {
 		return errSRNotEnabled
+	}
+	if rreq.RequestingDepID != "" {
+		// validate if requesting site is still part of site replication
+		var foundRequestor bool
+		for _, p := range c.state.Peers {
+			if p.DeploymentID == rreq.RequestingDepID {
+				foundRequestor = true
+				break
+			}
+		}
+		if !foundRequestor {
+			return errSRRequestorNotFound
+		}
 	}
 
 	ourName := ""

--- a/cmd/site-replication.go
+++ b/cmd/site-replication.go
@@ -2119,10 +2119,7 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 			errdID = dID
 		}
 	}
-	var remoteErr error // make sure to return err from remote if any
-	if errdID != "" {
-		remoteErr = errs[errdID]
-	}
+
 	// force local config to be cleared even if peers failed since the remote targets are deleted
 	// by now from the replication config and user intended to forcibly clear all site replication
 	if rreq.RemoveAll {
@@ -2157,16 +2154,17 @@ func (c *SiteReplicationSys) RemovePeerCluster(ctx context.Context, objectAPI Ob
 		return madmin.ReplicateRemoveStatus{
 			Status:    madmin.ReplicateRemoveStatusPartial,
 			ErrDetail: fmt.Sprintf("unable to save cluster-replication state on local: %v", err),
-		}, remoteErr
+		}, err
 	}
 
 	st = madmin.ReplicateRemoveStatus{
 		Status: madmin.ReplicateRemoveStatusSuccess,
 	}
-	if remoteErr != nil {
+	if errs[errdID] != nil {
 		st.Status = madmin.ReplicateRemoveStatusPartial
+		st.ErrDetail = errs[errdID].Error()
 	}
-	return st, remoteErr
+	return st, nil
 }
 
 // InternalRemoveReq - sends an unlink request to peer cluster to remove one or more sites

--- a/go.mod
+++ b/go.mod
@@ -49,7 +49,7 @@ require (
 	github.com/minio/dperf v0.4.2
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes v0.22.2
-	github.com/minio/madmin-go/v2 v2.0.2
+	github.com/minio/madmin-go/v2 v2.0.3
 	github.com/minio/minio-go/v7 v7.0.45
 	github.com/minio/pkg v1.5.8
 	github.com/minio/selfupdate v0.5.0

